processor/transformprocessor/internal/metrics/func_extract_sum_metric.go

```diff
@@ -0,0 +1,108 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package metrics // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/transformprocessor/internal/metrics"
+
+import (
+	"context"
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottlmetric"
+)
+
+type extractSumMetricArguments struct {
+	Monotonic bool      `ottlarg:"0"`
+	AggTemp   ottl.Enum `ottlarg:"1"`
+}
+
+func newExtractSumMetricFactory() ottl.Factory[ottlmetric.TransformContext] {
+	return ottl.NewFactory("extract_sum_metric", &extractSumMetricArguments{}, createExtractSumMetricFunction)
+}
+
+func createExtractSumMetricFunction(_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	args, ok := oArgs.(*extractSumMetricArguments)
+
+	if !ok {
+		return nil, fmt.Errorf("extractSumMetricFactory args must be of type *extractSumMetricArguments")
+	}
+
+	return extractSumMetric(args.Monotonic, args.AggTemp)
+}
+
+// this interface helps unify the logic for extracting data from different histogram types
+// all supported metric types' datapoints implement it
+type SumCountDataPoint interface {
+	Attributes() pcommon.Map
+	Sum() float64
+	Count() uint64
+	StartTimestamp() pcommon.Timestamp
+	Timestamp() pcommon.Timestamp
+}
+
+func extractSumMetric(monotonic bool, aggTempEnum ottl.Enum) (ottl.ExprFunc[ottlmetric.TransformContext], error) {
+	aggTemp := pmetric.AggregationTemporality(aggTempEnum)
+	switch aggTemp {
+	case pmetric.AggregationTemporalityDelta, pmetric.AggregationTemporalityCumulative:
+	default:
+		return nil, fmt.Errorf("unknown aggregation temporality: %s", aggTemp.String())
+	}
+
+	return func(_ context.Context, tCtx ottlmetric.TransformContext) (interface{}, error) {
+		metric := tCtx.GetMetric()
+		invalidMetricTypeError := fmt.Errorf("extract_sum_metric requires an input metric of type Histogram, ExponentialHistogram or Summary, got %s", metric.Type())
+
+		switch metric.Type() {
+		case pmetric.MetricTypeHistogram, pmetric.MetricTypeExponentialHistogram, pmetric.MetricTypeSummary:
+		default:
+			return nil, invalidMetricTypeError
+		}
+
+		sumMetric := tCtx.GetMetrics().AppendEmpty()
```

Comment:
Not sure where the easiest place to perform this check is, but I think it likely makes the most sense for us to not add a metric without any data points, in the case where all histogram data points don't contain a sum value.

Reply:
I changed this to create this metric independently, and only move it to the context's metric slice if the number of datapoints is greater than 0. This is much simpler than performing the check before doing any work, and not much more expensive.